File: solutions/80. Remove Duplicates from Sorted Array II.py
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        # time O(N), space O(n)
        ptr = 2
        for i in range(2, len(nums)):
            if nums[i] == nums[ptr-1] == nums[ptr-2]:
                continue
            nums[ptr] = nums[i]
            ptr += 1
        return ptr
        
       # #Recent attempt
        # for i in range(len(nums)-1, 1, -1):
        #     if nums[i] == nums[i-1] == nums[i-2]:
        #         nums.pop(i)
        # return len(nums)
    
#         ### Soln 2 - two pointers, optimal
#         j = 1
#         count = 1
#         for i in range(1, len(nums)):
#             if nums[i-1] == nums[i]:
#                 count += 1
#             else:
#                 count = 1
#             if count <= 2:
#                 nums[j] = nums[i]
#                 j += 1
#         return j

File: solutions/80. Remove Duplicates from Sorted Array II.cpp
class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        //2nd attempt, time O(N), space O(1)
        int count=1, j=1;
        for (int i=1; i<nums.size(); i++) {
            if (nums[i-1]==nums[i]) 
                count++;
            else {
                count = 1;
            }
            
            if (count <= 2) {
                nums[j++] = nums[i];
            }
        }
        return j;
        
//         // first attempt, time O(N), space O(N)
//         if (nums.size() == 0) return 0;
        
//         vector<pair<int, int>> record;
//         int temp = nums[0], count = 1;
//         for (int i = 1; i < nums.size(); i++){
//             if (temp == nums[i]) count++;
//             else {
//                 record.push_back(make_pair(temp,count));
//                 temp = nums[i];
//                 count = 1;
//             }
//         }
//         record.push_back(make_pair(temp,count));
//         int i = 0;
//         for (auto tp:record) {
//             if (tp.second > 2) tp.second = 2;
//             for (int j = 0; j < tp.second; j++){
//                 nums[i++] = tp.first;
//             }
//         }
//         return i;
    }
};
